readme.md
# Extract Colors

[![package version](https://badge.fury.io/js/extract-colors.svg)](https://www.npmjs.com/package/extract-colors)
[![npm min bundle size](https://img.shields.io/bundlephobia/min/extract-colors?style=flat&color=brightgreen)](https://bundlephobia.com/result?p=extract-colors)
[![npm gzip bundle size](https://img.shields.io/bundlephobia/minzip/extract-colors?style=flat&color=brightgreen)](https://bundlephobia.com/result?p=extract-colors)
![zero dependency](https://img.shields.io/badge/dependency-zero-brightgreen)
[![CI](https://github.com/Namide/extract-colors/workflows/CI/badge.svg)](https://github.com/Namide/extract-colors/actions)
[![code coverage](https://codecov.io/gh/Namide/extract-colors/branch/master/graph/badge.svg?token=80PUQ24PW5)](https://codecov.io/gh/Namide/extract-colors)
[![GNU GPL software License](https://img.shields.io/badge/license-GNU_GPL-brightgreen.svg)](#license)
[![Downloaded](https://img.shields.io/npm/dt/extract-colors)](https://www.npmjs.com/package/extract-colors)

Extract color palettes from images.  
Simple use, < 6ko minified, fast process and no dependencies for browser.  
Need image reader for node.js

[Website](https://extract-colors.namide.com/) | [Demo](https://extract-colors.namide.com/demo/) | [Guide](https://extract-colors.namide.com/guide/)

![3 examples of colors extraction](./doc/colors-2.jpg)

## Requirements

### Browsers

- Firefox: 29+
- Chrome: 33+
- Edge: 12+
- Opera: 19+
- Safari: 8+
- Webview Android: 4.4.3+
- Samsung Internet: 2.0+
- ~~Internet Explorer~~


### Node

- Node.js: 6.0+


## Install

### For browser

```bash
npm install --save extract-colors
```


### For node.js

Need to install an ImageData extractor like `get-pixels`

```bash
npm install --save extract-colors get-pixels
```


## Usage

### Browser example

```js
import { extractColors } from 'extract-colors'

const src = 'my-image.jpg'

extractColors(src)
  .then(console.log)
  .catch(console.error)
```

> You can use different types for `src` param (`String` for a path of image, `HTMLImageElement` or `ImageData`).


### Node.js example

```js
const path = require('path')
const getPixels = require("get-pixels")
const { extractColors } = require('extract-colors')

const src = path.join(__dirname, './my-image.jpg')

getPixels(src, (err, pixels) => {
  if(!err) {
    const data = [...pixels.data]
    const width = Math.round(Math.sqrt(data.length / 4))
    const height = width

    extractColors({ data, width, height })
      .then(console.log)
      .catch(console.log)
  }
})
```

> This example use `get-pixels` but you can change the lib.
> Just send and ImageData object to `extractColors(imageData)`.


### ExtractorOptions

```js
const options = {
  pixels: 64000,
  distance: 0.22,
  colorValidator: (red, green, blue, alpha = 255) => alpha > 250,
  saturationDistance: 0.2,
  lightnessDistance: 0.2,
  hueDistance: 0.083333333
}

extractColors(src, options)
  .then(console.log)
  .catch(console.error)
```

**pixels**  
_Total pixel number of the resized picture for calculation_  
Type: `Integer`  
Default: `64000`  

**distance**  
_From 0 to 1 is the color distance to not have near colors (1 distance is between white and black)_  
Type: `Number`  
Default: `0.22`  

**colorValidator**  
_Test function to enable only some colors_  
Type: `Function`  
Default: `(red, green, blue, alpha = 255) => alpha > 250`

**crossOrigin**  
_Only for browser, can be 'Anonymous' to avoid client side CORS_
_(the server side images need authorizations too)_  
Type: `String`  
Default: `null`

**saturationDistance**  
_Minimum saturation value between two colors otherwise the colors will be merged (from 0 to 1)_
Type: `String`  
Default: `0.2`

**lightnessDistance**  
_Minimum lightness value between two colors otherwise the colors will be merged (from 0 to 1)_
Type: `String`  
Default: `0.2`

**hueDistance**  
_Minimum hue value between two colors otherwise the colors will be merged (from 0 to 1)_
Type: `String`  
Default: `0.083333333`


## Return of the promise

Array of colors with the followed properties:

```js
[
  {
    hex: "#858409",​​
    red: 133,​​
    green: 132,​​
    blue: 9,​​
    hue: 0.16532258064516128,​​
    intensity: 0.4862745098039216,​​
    lightness: 0.2784313725490196,​​
    saturation: 0.8732394366197184,
    area: 0.0004
  },
  ...
]
```

| Field | Example | Type | Description |
|---|---|---|---|
| hex | #858409 | String | color in hexadecimal string |
| red | 133 | Integer | red canal from 0 to 255 |
| green | 132 | Integer | green canal from 0 to 255 |
| blue | 9 | Integer | blue canal from 0 to 255 |
| hue | 0.1653 | Number | color tone from 0 to 1 |
| intensity | 0.4862 | Number | color intensity from 0 to 1 |
| lightness | 0.2784 | Number | color lightness from 0 to 1 |
| saturation | 0.8732 | Number | color saturation from 0 to 1 |
| area | 0.0004 | Number | area of the color and his neighbouring colors from 0 to 1 |


## License

Copyright (C) 2019  Damien Doussaud

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.


changelog.md
# Changelog


## v4.0.2 _(2023-09-06)_

- Improve bundle size


## v4.0.1 _(2023-09-05)_

- Update filename version for better node.js module usage


## v4.0.0 _(2023-09-03)_

- Remove errors and warnings from production build
- Fix color groups keys
- Optimize tree performances
- Remove splitPower option
- Change default values
- Fix area values


## v3.0.0 _(2023-04-13)_

- Remove errors and warnings from production build
- Fix color groups keys
- Optimize tree performances
- Remove `splitPower` option
- Change default values
- Fix area values


## v2.0.6 _(2023-04-06)_

- Fix medium key overlap


## v2.0.5 _(2023-02-07)_

- Remove default import from doc


## v2.0.4 _(2023-02-07)_

- Improve bundle size


## v2.0.3 _(2023-12-04)_

- Rewriting in Typescript
- Add HSL conversion to sort and filter finals colors with new arguments (saturationDistance, lightnessDistance, hueDistance)
- Remove canvas dependency to NodeJs version
- Remove src and Image inputs from NodeJs version (only use ImageData)


sample/script.js
import extractColors from '../src/extractColorsBrowser'

let realState = 0
let imgs = []

const options = {
  pixels: 10000,
  distance: 0.2,
  saturationImportance: 0.2,
  splitPower: 10,
  uploadFile () {
    imgs = []
    const input = document.getElementById('inputFile')
    input.onchange = function ({ target }) {
      [...target.files].forEach(file => {
        imgs.push(URL.createObjectURL(file))
      })
      process()
    }
    input.click()
  },
  randomFile () {
    imgs = []
    randomImgs()
    process()
  }
}

// Add image + colors in DOM
const resetDisplay = () => {

  const list = document.body.querySelector('.list')
  list.innerHTML = ''

  imgs.forEach(src => {
    const div = document.createElement('div')
    div.classList.add('block')

    // display image
    const image = new Image()
    image.src = src
    image.height = 200
    image.style.width = 'auto'
    div.appendChild(image)

    const canvas = document.createElement('canvas')
    canvas.width = 20
    canvas.height = image.height
    div.appendChild(canvas)

    list.appendChild(div)  
  })
}

const resetLog = () => {
  document.body.querySelector('.log').innerHTML = ''
}

const displayLog = (infos) => {

  const log = document.body.querySelector('.log')
  const div = document.createElement('div')
  div.innerHTML = infos
  log.prepend(div)
}

const displayImg = ({ colors, index, state, initTime }) => {

  if (state !== realState) {
    return false
  }

  displayLog(' ')
  displayLog('time: ' + (Date.now() - initTime) + 'ms' )
  displayLog('colors: ' + colors.map(color => {
    return `<span class="square" style="color:${color.hex}">▮</span>`
  }).join(''))
  displayLog(`<strong>Image ${index + 1}</strong>`)
  
  const div = document.body.querySelectorAll('.list .block')[index]
  const canvas = div.querySelector('canvas')

  // display colors
  {
    const context = canvas.getContext('2d')

    const imageData = context.createImageData(canvas.width, canvas.height)
    const pixels = imageData.data

    /* eslint no-labels: "off" */
    loop: for (let r = 0; r < imageData.height; r++) {
      for (let c = 0; c < imageData.width; c++) {
        const i = Math.floor(r / canvas.width)
        const color = colors[i]

        if (!color) { break loop }

        const pos = (r * (imageData.width * 4)) + (c * 4)
        pixels[pos + 0] = color.red
        pixels[pos + 1] = color.green
        pixels[pos + 2] = color.blue
        pixels[pos + 3] = 255
      }
    }

    context.putImageData(imageData, 0, 0)
  }

  return true
}

const update = (data) => {
  console.log(data, options)
  process()
}

const randomImgs = ({ random = true, directory = false } = {}) => {
  
  if (random) {
    // Some random tests
    ['moon', 'water', 'sea', 'colors', 'sky'].forEach(seed => {
      const index = Math.round(Math.random() * 20) + 5
      imgs.push(`https://loremflickr.com/320/240/${seed}/?lock=${index}`)
    })
  } else if (directory) {
    // Your tests in the img directory
    Object.values(require('./img/*.jpg')).forEach(imgSrc => {
      imgs.push(imgSrc)
    })
  }
}

const process = () => {
  realState++
  let state = realState

  resetLog()
  resetDisplay()

  let promise
  imgs.forEach((src, index) => {
    if (!promise) {
      const initTime = Date.now()
      promise = extractColors(src, { ...options, crossOrigin: 'anonymous' })
        .then(colors => displayImg({ colors, src, index, state, initTime }))
    } else {
      promise
        .then(() => {
          const initTime = Date.now()
          return extractColors(src, { ...options, crossOrigin: 'anonymous' })
            .then(colors => displayImg({ colors, src, index, state, initTime }))
        })
    }
  })

  if (promise) {
    promise.catch(console.log)
  }
}


const gui = new dat.GUI()
gui.add(options, 'pixels', 1).step(1).name("pixels").onFinishChange(update)
gui.add(options, 'distance', 0, 1).name("distance").onFinishChange(update)
gui.add(options, 'saturationImportance', 0, 1).name("saturationImportance").onFinishChange(update)
gui.add(options, 'splitPower', 2, 16).name("splitPower").onFinishChange(update)

gui.add(options, 'uploadFile').name('Upload images')
gui.add(options, 'randomFile').name('5 random images')

options.randomFile()
